File: src/main/java/com/atsuishio/superbwarfare/init/ModMenuTypes.java
package com.atsuishio.superbwarfare.init;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.inventory.menu.BlueprintResearchTableMenu;
import com.atsuishio.superbwarfare.inventory.menu.MediumVehicleContainerMenu;
import com.atsuishio.superbwarfare.menu.*;
import net.minecraft.world.inventory.MenuType;
import net.minecraftforge.common.extensions.IForgeMenuType;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;
import net.minecraftforge.registries.RegistryObject;

public class ModMenuTypes {

    public static final DeferredRegister<MenuType<?>> REGISTRY = DeferredRegister.create(ForgeRegistries.MENU_TYPES, Mod.MODID);

    public static final RegistryObject<MenuType<ReforgingTableMenu>> REFORGING_TABLE_MENU =
            REGISTRY.register("reforging_table_menu",
                    () -> IForgeMenuType.create((windowId, inv, data) -> new ReforgingTableMenu(windowId, inv)));
    public static final RegistryObject<MenuType<ChargingStationMenu>> CHARGING_STATION_MENU =
            REGISTRY.register("charging_station_menu",
                    () -> IForgeMenuType.create((windowId, inv, data) -> new ChargingStationMenu(windowId, inv)));

    public static final RegistryObject<MenuType<MediumVehicleContainerMenu>> MEDIUM_VEHICLE_CONTAINER_MENU =
            REGISTRY.register("medium_vehicle_container", () -> MediumVehicleContainerMenu.TYPE);

    public static final RegistryObject<MenuType<SuperbItemInterfaceMenu>> SUPERB_ITEM_INTERFACE_MENU =
            REGISTRY.register("superb_item_interface_menu",
                    () -> IForgeMenuType.create((windowId, inv, data) -> new SuperbItemInterfaceMenu(windowId, inv)));
    public static final RegistryObject<MenuType<FuMO25Menu>> FUMO_25_MENU =
            REGISTRY.register("fumo_25_menu",
                    () -> IForgeMenuType.create((windowId, inv, data) -> new FuMO25Menu(windowId, inv)));
    public static final RegistryObject<MenuType<VehicleAssemblingMenu>> VEHICLE_ASSEMBLING_MENU =
            REGISTRY.register("vehicle_assembling_menu",
                    () -> IForgeMenuType.create((windowId, inv, data) -> new VehicleAssemblingMenu(windowId, inv)));
    public static final RegistryObject<MenuType<BlueprintResearchTableMenu>> BLUEPRINT_RESEARCH_TABLE =
            REGISTRY.register("blueprint_research_table_menu",
                    () -> IForgeMenuType.create(((windowId, inv, data) -> new BlueprintResearchTableMenu(windowId, inv))));
}

File: src/main/java/com/atsuishio/superbwarfare/init/ModMenuTypes.kt
package com.atsuishio.superbwarfare.init

import com.atsuishio.superbwarfare.Mod
import com.atsuishio.superbwarfare.inventory.menu.BlueprintResearchTableMenu
import com.atsuishio.superbwarfare.inventory.menu.MediumVehicleContainerMenu
import com.atsuishio.superbwarfare.inventory.menu.SmallVehicleContainerMenu
import com.atsuishio.superbwarfare.menu.*
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.MenuType
import net.neoforged.neoforge.common.extensions.IMenuTypeExtension
import net.neoforged.neoforge.network.IContainerFactory
import net.neoforged.neoforge.registries.DeferredHolder
import net.neoforged.neoforge.registries.DeferredRegister
import java.util.function.Supplier

object ModMenuTypes {
    @JvmField
    val REGISTRY: DeferredRegister<MenuType<*>> = DeferredRegister.create(BuiltInRegistries.MENU, Mod.MODID)

    private fun <T : AbstractContainerMenu> register(
        name: String, factory: IContainerFactory<T>
    ): DeferredHolder<MenuType<*>, MenuType<T>> =
        REGISTRY.register(name, Supplier { IMenuTypeExtension.create(factory) })

    private fun <T : AbstractContainerMenu> register(
        name: String, factory: () -> MenuType<T>
    ): DeferredHolder<MenuType<*>, MenuType<T>> =
        REGISTRY.register(name, factory)

    @JvmField
    val REFORGING_TABLE_MENU =
        register("reforging_table_menu") { windowId, inv, _ -> ReforgingTableMenu(windowId, inv) }

    @JvmField
    val CHARGING_STATION_MENU =
        register("charging_station_menu") { windowId, inv, _ -> ChargingStationMenu(windowId, inv) }

    @JvmField
    val SMALL_VEHICLE_CONTAINER_MENU = register("small_vehicle_container") { SmallVehicleContainerMenu.TYPE }

    @JvmField
    val MEDIUM_VEHICLE_CONTAINER_MENU = register("medium_vehicle_container") { MediumVehicleContainerMenu.TYPE }

    @JvmField
    val SUPERB_ITEM_INTERFACE_MENU =
        register("superb_item_interface_menu") { windowId, inv, _ -> SuperbItemInterfaceMenu(windowId, inv) }

    @JvmField
    val FUMO_25_MENU = register("fumo_25_menu") { windowId, inv, _ -> FuMO25Menu(windowId, inv) }

    @JvmField
    val VEHICLE_ASSEMBLING_MENU =
        register("vehicle_assembling_menu") { windowId, inv, _ -> VehicleAssemblingMenu(windowId, inv) }

    @JvmField
    val BLUEPRINT_RESEARCH_TABLE =
        register("blueprint_research_table_menu") { windowId, inv, _ -> BlueprintResearchTableMenu(windowId, inv) }
}
